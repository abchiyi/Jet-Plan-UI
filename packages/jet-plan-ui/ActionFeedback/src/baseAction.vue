<script lang="ts">
import { h, defineComponent } from "vue";
const name = "j-base-action";

export default defineComponent({
  name: name,
  props: {
    tag: { default: "div" },
  },
  data() {
    return {
      touchIsStarted: false,
    };
  },
  emits: ["active_from", "active_to"],
  methods: {
    activeFrom(event: Event) {
      this.$emit("active_from", event);
    },
    activeTo(event: Event) {
      this.$emit("active_to", event);
    },
    handlerMouseEvent(event: Event) {
      if (this.touchIsStarted) return;
      switch (event.type) {
        case "mousedown":
          this.activeFrom(event);
          break;
        case "mouseup":
          this.activeTo(event);
          break;
        case "mouseleave":
          this.activeTo(event);
          break;
      }
    },

    handlerTouchEvent(event: Event) {
      switch (true) {
        case "touchstart" === event.type:
          this.touchIsStarted = true;
          this.activeFrom(event);
          // this.hoverFrom(event);

          break;

        case ["touchend", "touchcancel"].includes(event.type):
          this.activeTo(event);
          // this.hoverTo(event);
          this.touchIsStarted = false;
          break;
      }
    },
  },
  render() {
    return h(
      this.tag,
      {
        class: [name],
        // Mouse
        onmousedown: this.handlerMouseEvent,
        onmouseup: this.handlerMouseEvent,
        onmouseleave: this.handlerMouseEvent,
        // Touch
        ontouchstart: this.handlerTouchEvent,
        ontouchcancel: this.handlerTouchEvent,
        ontouchend: this.handlerTouchEvent,
      },
      {
        default: () => [this.$slots.default?.()],
      }
    );
  },
});
</script>
