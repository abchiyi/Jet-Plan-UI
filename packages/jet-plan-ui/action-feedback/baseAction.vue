<script>
import { h } from 'vue';
import { propInitBoolean } from '../tool/lib';
const name = 'j-base-action';
export default {
    name: name,
    props: {
        action: propInitBoolean(false),
        hover: propInitBoolean(false),
        focus: propInitBoolean(false),
    },
    computed: {
        class() {
            return [
                this.action ? 'action' : null,
                this.hover ? 'hover' : null,
                this.focus ? 'focus' : null,
            ];
        },
    },
    render() {
        return h('div', {
            class: [name, ...this.class],
        });
    },
};
</script>

<style>
</style>
