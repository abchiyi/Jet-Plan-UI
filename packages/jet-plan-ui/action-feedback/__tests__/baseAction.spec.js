import {
    mount,
    shallowMount
} from '@vue/test-utils'

import baseAction from '../baseAction.vue'
import {
    getClass,
    getLastEvent
} from '@test/tools'
import {
    get
} from 'jquery'


describe('Base ActionFeedback', () => {

    it('Node Map', async () => {
        const wrapper = mount(baseAction)

        expect(wrapper.html()).toMatchSnapshot()
    })

    it('Class Map', async () => {
        const wrapper = shallowMount(baseAction)

        // 类名匹配组件名
        expect(getClass(wrapper)).toContain(
            baseAction.name
        )

    })

    it("Prop tag", () => {
        const wrapper = shallowMount(baseAction, {
            propsData: {
                tag: 'p'
            }
        })

        expect(wrapper.element.tagName).toEqual('P')
    })

    it('Actions', async () => {
        const wrapper = mount(baseAction)

        wrapper.trigger('mousedown')
        expect(
            getLastEvent(wrapper, 'active_from').active
        ).toBeTruthy()

        wrapper.trigger('mouseup')
        expect(
            getLastEvent(wrapper, 'active_to').active
        ).not.toBeTruthy()

        // 触摸事件监听
        wrapper.vm.handlerTouchEvent({
            type: 'touchstart'
        })
        expect(
            getLastEvent(wrapper, 'active_from').active
        ).toBeTruthy()

        wrapper.vm.handlerTouchEvent({
            type: 'touchend'
        })
        expect(
            getLastEvent(wrapper, 'active_to')
        ).toEqual({
            active: false,
            event: {
                type: 'touchend'
            }
        })
        wrapper.vm.handlerTouchEvent({
            type: 'touchcancel'
        })
        expect(
            getLastEvent(wrapper, 'active_to')
        ).toEqual({
            active: false,
            event: {
                type: 'touchcancel'
            }
        })
    })
    it('Actions on touch', async () => {
        const wrapper = mount(baseAction)

        wrapper.vm.isTouch = true

        wrapper.trigger('mousedown')
        expect(wrapper.emitted().active_from).not.toBeTruthy()

        wrapper.trigger('mouseup')
        expect(wrapper.emitted().active_to).not.toBeTruthy()
    })

    it("Hover", async () => {
        const wrapper = mount(baseAction)

        wrapper.trigger('mouseenter')
        expect(
            getLastEvent(wrapper, 'hover_from').active
        ).toBeTruthy()

        wrapper.trigger('mouseleave')
        expect(getLastEvent(wrapper, 'hover_to').active).toBe(false)
    })

    it('Hover on touch', async () => {
        const wrapper = mount(baseAction)

        wrapper.vm.handlerTouchOver('')
        expect(wrapper.emitted().active_to)
        expect(wrapper.emitted().hover_to)

    })

})
