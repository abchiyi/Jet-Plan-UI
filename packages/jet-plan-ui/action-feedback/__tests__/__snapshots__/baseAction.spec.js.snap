// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Base ActionFeedback Node Map 1`] = `<div class="j-base-action"></div>`;
